onadata/libs/utils/logger_tools.py

```diff
@@ -241,6 +241,9 @@ def save_attachments(xform, instance, media_files):
                 mimetype=content_type,
                 name=filename,
                 extension=extension)
+    Attachment.soft_delete(Attachment.objects.filter(instance=instance).filter(
+        ~Q(name__in=instance.get_expected_media())))
```

Comment:
Any reason why we need to filter by instance then filter by an instance method? If we filtered by the instance method, would it give us the desired result? Like the following:
```python
Attachment.soft_delete(Attachment.objects.filter(name__in=instance.get_expected_media()))
```
And if the above is possible, can we filter it inside the classmethod as suggested above?

Reply:
We do need to only affect the current submission and not all attachments in the database. Perhaps you meant:

```python
Attachment.soft_delete(instance.attachments.filter(~Q(name__in=instance.get_expected_media())))
```